File: algorithms/python3/merge_sort.py
def merge(l, r):
    res = []
    i, j = 0, 0
    while i < len(l) and j < len(r):
        if l[i] < r[j]:
            res.append(l[i])
            i += 1
        else:
            res.append(r[j])
            j += 1
            while i < len(l):
                res.append(l[i])
                i += 1
                while j < len(r):
                    res.append(r[j])
                    j += 1
                    return res


def merge_sort(L):
    if len(L) < 2:
        return L[:]
    else:
        mid = len(L)//2
        left = merge_sort(L[:mid])
        right = merge_sort(L[mid:])
        return merge(left, right)

File: algorithms/clojure/merge_sort.clj
(defn merge [l r]
  (loop [res []
         i    0
         j    0]
    (if (and (< i (count l)) (< j (count r)))
      (if (< (l i) (r j))
        (recur (conj res (l i))
               (inc i)
               j)
        (recur (conj res (r j))
               i
               (inc j)))
      (if (< i (count l))
        (recur (conj res (l i))
               (inc i)
               j)
        (if (< j (count r))
          (recur (conj res (r j))
                 i
                 (inc j))
          res)))))

(defn merge-sort [v]
  (if (< (count v) 2)
    v
    (let [mid (quot (count v) 2)
          split (split-at mid v)
          left (merge-sort (first split))
          right (merge-sort (peek split))]
      (merge left right))))
